coderd/keyrotate/rotate.go

```diff
@@ -0,0 +1,287 @@
+	// The starts at of the new key is the expiration of the old key. We set the deletes_at of the old key to a little over
+	// an hour after its set to expire so there should plenty
+	// of time for the new key to enter rotation.
+	newStartsAt := key.ExpiresAt(k.keyDuration)
+
+	newKey, err := k.insertNewKey(ctx, tx, key.Feature, newStartsAt, key.Sequence+1)
+	if err != nil {
+		return nil, xerrors.Errorf("insert new key: %w", err)
+	}
+
+	// Set old key's deletes_at to an hour + however long the token
+	// for this feature is expected to be valid for. This should
+	// allow for sufficient time for the new key to propagate to
+	// dependent services (i.e. Workspace Proxies).
+	deletesAt := newStartsAt.Add(time.Hour).Add(tokenDuration(key.Feature))
+
+	updatedKey, err := tx.UpdateCryptoKeyDeletesAt(ctx, database.UpdateCryptoKeyDeletesAtParams{
+		Feature:  key.Feature,
+		Sequence: key.Sequence,
+		DeletesAt: sql.NullTime{
+			Time:  deletesAt.UTC(),
+			Valid: true,
+		},
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("update old key's deletes_at: %w", err)
+	}
+
+	return []database.CryptoKey{updatedKey, newKey}, nil
+}
+
+func (k *Rotator) Close() {
+	k.ticker.Stop()
+}
+
+func generateNewSecret(feature database.CryptoKeyFeature) (string, error) {
+	switch feature {
+	case database.CryptoKeyFeatureWorkspaceApps:
+		return generateKey(96)
+	case database.CryptoKeyFeatureOidcConvert:
+		return generateKey(32)
+	case database.CryptoKeyFeatureTailnetResume:
+		return generateKey(64)
+	}
+	return "", xerrors.Errorf("unknown feature: %s", feature)
+}
+
+func generateKey(length int) (string, error) {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", xerrors.Errorf("rand read: %w", err)
+	}
+	return hex.EncodeToString(b), nil
+}
+
+func tokenDuration(feature database.CryptoKeyFeature) time.Duration {
+	switch feature {
+	case database.CryptoKeyFeatureWorkspaceApps:
+		return WorkspaceAppsTokenDuration
+	case database.CryptoKeyFeatureOidcConvert:
+		return OIDCConvertTokenDuration
+	case database.CryptoKeyFeatureTailnetResume:
+		return TailnetResumeTokenDuration
+	default:
+		return 0
+	}
+}
+
+func shouldDeleteKey(key database.CryptoKey, now time.Time) bool {
+	return key.DeletesAt.Valid && !now.Before(key.DeletesAt.Time.UTC())
+}
+
+func shouldRotateKey(key database.CryptoKey, keyDuration time.Duration, now time.Time) bool {
+	// If deletes_at is set, we've already inserted a key.
+	if key.DeletesAt.Valid {
+		return false
+	}
+	expirationTime := key.ExpiresAt(keyDuration)
+	return !now.Add(time.Hour).UTC().Before(expirationTime)
+}
+
+func keysByFeature(keys []database.CryptoKey, features []database.CryptoKeyFeature) map[database.CryptoKeyFeature][]database.CryptoKey {
+	m := map[database.CryptoKeyFeature][]database.CryptoKey{}
+	for _, feature := range features {
+		m[feature] = []database.CryptoKey{}
+	}
+	for _, key := range keys {
+		m[key.Feature] = append(m[key.Feature], key)
```

Comment:
We should check here that it's a feature we know about, otherwise we could attempt to rotate a key for a feature that we don't know the appropriate duration of.

Reply:
The feature type is an enum in the database so that shouldn't be possible. I'm guessing you mean as a safeguard for the case where someone adds an enum and neglects to update the logic?

Reply:
I was thinking an older version of Coderd running against a newer version of the database schema.

Reply:
I wasn't aware you could successfully get `coderd` to start if it detected the migrations were out of sync.